source/common/thread_local/thread_local_impl.cc

```diff
@@ -91,6 +91,20 @@ void InstanceImpl::runOnAllThreads(Event::PostCb cb) {
   cb();
 }
 
+void InstanceImpl::runOnAllThreadsWithBarrier(Event::PostCb cb, Event::PostCb main_callback) {
+  ASSERT(std::this_thread::get_id() == main_thread_id_);
+  ASSERT(!shutdown_);
+  worker_counter_ = registered_threads_.size();
```

Comment:
This is not safe for concurrent uses of this function. Instead of using a member variable, allocate an atomic inside a shared_ptr, and capture it in the posted lambdas. 

Also, this new function needs explicit tests in thread_local_impl_test

Reply:
would it be used concurrently since it is allowed to be called only from main thread? Any way I will try with shared_ptr as it seems better compared to member variable.

Reply:
I have changed it to use shared_ptr. I came across this http://floating.io/2017/07/lambda-shared_ptr-memory-leak/ - Should we worry about this?

Reply:
reg tests, I will add once the implementation comes to a sane stage.